go/saltpack/decrypt.go

```diff
@@ -249,13 +261,16 @@ func (ds *decryptStream) processEncryptionHeader(hdr *EncryptionHeader) error {
 		if longLivedSenderKey == nil {
 			return ErrNoSenderKey
 		}
-		ds.tagKey = secretKey.Precompute(longLivedSenderKey)
 		ds.mki.SenderKey = longLivedSenderKey
 	} else {
 		ds.mki.SenderIsAnon = true
 		ds.mki.SenderKey = ephemeralKey
 	}
 
+	// Compute the MAC key.
+	macKeyBox := secretKey.Box(ds.mki.SenderKey, nonceForMACKeyBox(ds.headerHash[:]), make([]byte, 32))
+	ds.macKey = macKeyBox[16:48]
```

Comment:
Will do. Regarding the second idea, `crypto_auth_KEYBYTES` is defined to be 32. Even though our own HMAC functions in Go and Python can accept arbitrary-sized keys, I want to stay compatible with the NaCl interface.

One option would be to encrypt 16 zero bytes instead of 32 zero bytes. Then the resulting box would be exactly 32 bytes, and in theory we could use the whole thing. I'm against doing that for a couple reasons:
- Using Poly1305 output as though it were XSalsa20 output feels squicky. (Even though the construction of `crypto_box` means that it really *is* XSalsa20 output.)
- If we stick to the payload part of the `crypto_box`, that means that in theory someone could use `crypto_stream` to skip the auth step entirely and compute the MAC key more efficiently. (It only saves a small, constant amount of time, so I'm not sure I actually expect anyone to make this optimization, but who knows.)